File: dataset/human/java/file_9698.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.InputMismatchException;
import java.util.StringTokenizer;

public class Main {

	static class BinaryTrees {
		public void binarytrees (InputReader ir) {
			
			int n = ir.nextInt();
			Node[] node = new Node[n];
			for (int i = 0; i < n; i++) {
				node[i] = new Node();
			}
			
			for (int i = 0; i < n; i++) {
				int id = ir.nextInt();
				node[id].left = ir.nextInt();
				node[id].right = ir.nextInt();
				if (node[id].left != -1 && node[id].right != -1) {
					node[id].degree = 2;
				} else if (node[id].left != -1 || node[id].right != -1) node[id].degree = 1;
			}
			
			Parent (node, n);
			Sibling (node, n);
			
			for (int i = 0; i < n / 2; i++) {
				Depth (node, n);
				Height (node, n);
			}
			
			StringBuilder sb = new StringBuilder();
			
			for (int i = 0; i < n; i++) {
				sb.append("node ").append(Integer.toString(i))
				.append(": parent = ").append(node[i].parent)
				.append(", sibling = ").append(node[i].sibling)
				.append(", degree = ").append(node[i].degree)
				.append(", depth = ").append(node[i].depth)
				.append(", height = ").append(node[i].height);
				
				if (node[i].parent == -1) {
					sb.append(", root\n");
				} else if (node[i].degree > 0) {
					sb.append(", internal node\n");
				} else {
					sb.append(", leaf\n");
				}
			}			
			System.out.print(sb);
		}
		
		static void Parent (Node[] node, int n) {
			
			for (int i = 0; i < n; i++) {
				if (node[i].degree == 2) {
					node[node[i].left].parent = i;
					node[node[i].right].parent = i;
				} else if (node[i].degree == 1) {
					if (node[i].left != -1) node[node[i].left].parent = i;
					else node[node[i].right].parent = i;
				}
			}
			
		}
	
		static void Sibling (Node[] node, int n) {
			
			for (int i = 0; i < n; i++) {
				if ( node[i].parent != -1) {
					if (node[node[i].parent].left == i && node[node[i].parent].right != -1) {
						node[i].sibling = node[node[i].parent].right;
					} else if (node[node[i].parent].right == i && node[node[i].parent].left != -1) {
						node[i].sibling = node[node[i].parent].left;
					}
							
				}
			}
			
		}

		static void Depth (Node[] node, int n) {
			
			for (int i = 0; i < n; i++) {
				if (node[i].degree == 2) {
					node[node[i].left].depth = node[i].depth + 1;
					node[node[i].right].depth = node[i].depth + 1;
				} else if (node[i].left != -1) {
					node[node[i].left].depth = node[i].depth + 1;
				} else if (node[i].right != -1) {
					node[node[i].right].depth = node[i].depth + 1;
				}
			}
		}
		
		static void Height (Node[] node, int n) {
			
			for (int i = 0; i < n; i++) {
				if (node[i].degree == 2) {
					node[i].height = Math.max(node[node[i].left].height + 1, node[node[i].right].height + 1);
				} else if (node[i].left != -1) {
					node[i].height = node[node[i].left].height + 1;
				} else if(node[i].right != -1) {
					node[i].height = node[node[i].right].height + 1;
				} 
			}
			
		}
	}
	
	static class Node {
		int parent;
		int sibling;
		int degree;
		int left;
		int right;
		int height;
		int depth;
		public Node() {
			parent = -1;
			sibling = -1;
			degree = 0;
 			left = -1;
			right = -1;
			depth = 0;
			
		}
		
	}
	
	static class InputReader {
		BufferedReader br;
		StringTokenizer st;
		
		public String nextString() {
			
			 while (!st.hasMoreTokens()) {
	                try {
	                    st = new StringTokenizer(br.readLine(), " ");
	                } catch (IOException e) {
	                    throw new InputMismatchException();
	                }
	            }
			return st.nextToken();
		}
		
		public int nextInt() {
			return Integer.parseInt(nextString());
		}
					
		public InputReader (InputStream inputStream) {
			br = new BufferedReader (new InputStreamReader (inputStream));
			st = new StringTokenizer ("");
		}
		
	}
	
	public static void main (String[] args) {
		InputStream inputStream = System.in;
		InputReader in = new InputReader (inputStream);
		BinaryTrees obj = new BinaryTrees();
		obj.binarytrees(in);
	}
}

File: dataset/human/c/file_9698.c
#include<stdio.h>
#define MAX 100005
#define NIL -1

struct NODE{
  int parent,left,right;
};
struct NODE T[MAX];
int n,D[MAX];

int getDepth(int u){
  int d = 0;
  while(T[u].parent != NIL){
    u = T[u].parent;
    d++;
  }
  return d;
}

void printChildren(int u){
  int c;
  c = T[u].left;
  printf("[");
  while(c != NIL){
    printf("%d",c);
    c = T[c].right;
    if(c != NIL)printf(", ");
  }
  printf("]\n");
}

int main(){
  int i,j,d,v,c,l,r;
  
  scanf("%d",&n);
  
  for(i = 0 ; i < n ; i++){
    T[i].parent = NIL;
    T[i].left = NIL;
    T[i].right = NIL;
  }
  
  for(i = 0 ; i < n ; i++){
    scanf("%d",&v);
    scanf("%d",&d);
    for(j = 0 ; j < d ; j++){
      scanf("%d",&c);
      if(j == 0)T[v].left = c;
      else T[l].right = c;
      l = c;
      T[c].parent = v;
    }
  }
  
  for(i = 0 ; i < n ; i++){
    if(T[i].parent == NIL)r = i;
  }

  for(i = 0 ; i < n ; i++){
    printf("node %d: ",i);
    printf("parent = %d, ",T[i].parent);
    printf("depth = %d, ",getDepth(i));
    if(T[i].parent == NIL)printf("root, ");
    else if(T[i].left == NIL)printf("leaf, ");
    else printf("internal node, ");
    printChildren(i);    
  }
  
  return 0;
}
